### PR TITLE
Streaming body parser: Multipart

### PR DESCRIPTION
**Pull request by WyriHaximus**
Supersedes / closes #72

### CODE REVIEW

File: src/StreamingBodyParser/MultipartParser.php [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    protected $onDataCallable;
+
+    public function __construct(RequestInterface $request)
+    {
+        $this->onDataCallable = array($this, 'onData');
+        $deferred = new Deferred(function () {
+            $this->body->removeListener('data', $this->onDataCallable);
+            $this->body->close();
+        });
+        $this->promise = $deferred->promise();
```
Comment (jsor) [inline src/StreamingBodyParser/MultipartParser.php]: Is the promise property really needed? Looks like it's only used in `MultipartParser::cancel()`. I think this is unecessary and 

```php
$this->body->removeListener('data', $this->onDataCallable);
$this->body->close();
``` 
can be just called directly there.

File: src/StreamingBodyParser/MultipartParser.php [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+    public function cancel()
+    {
+        $this->promise->cancel();
+    }
+
+    /**
+     * @internal
+     */
+    public function stripTrainingEOL($chunk)
```
Comment (jsor) [inline src/StreamingBodyParser/MultipartParser.php]: Typo: `stripTrainingEOL` -> `stripTrailingEOL`

File: src/StreamingBodyParser/MultipartParser.php (hunk 1 above)
Comment (WyriHaximus) [inline src/StreamingBodyParser/MultipartParser.php]: Not really, fixed it :+1:

**Comment by andig** [comment]
@WyriHaximus superseded by #223?

**Comment by WyriHaximus** [comment]
Superseded by #220 or #223

**Comment by docteurklein** [comment]
I'm sorry to necrobump, but how is #220 or #223 replacing this PR ?
Maybe I miss something, but it seemed to me that this PR would allow huge file uploads without actually filling the whole memory with the request body, no?

**Comment by clue** [comment]
@docteurklein No worries, I think this is a very relevant question :+1: 

As per https://github.com/reactphp/http/issues/28#issuecomment-292778882 and https://github.com/reactphp/http/issues/105 and we've decided to implement full PSR-7 support for v0.8.0 first. This includes form post handling and support for file uploads. We believe that this fulfills most use cases and as such resolves this issue for most users of this library.

While we have added relevant documentation and tests to our project for its limitations, I agree that it would be nice to *also* have streaming upload functionality built-in. That being said, I'm not aware of anybody currently working on this.

If you want to look into this (or want to sponsor this feature), then go for it! :shipit: Reach out if you have any questions!

**Comment by omar391** [comment]
@WyriHaximus or @clue sorry for bumping you, is this feature aka streaming upload implemented in the latest http lib?

**Comment by SimonFrings** [comment]
@omar391, yes this is implemented (with PSR-7 support), but there are limitations. @clue wrote a detailed answer in https://github.com/reactphp/http/issues/428#issuecomment-1030585707 , this should answer everything related to this topic :)

Hope this helps :+1:

**Comment by omar391** [comment]
@SimonFrings Thank you for the reference. I guess I need to wait for a capable streaming implementation. 🤔

BTW, can you please refer me to an implementation of multi part upload client?

Thank you!

**Comment by SimonFrings** [comment]
> BTW, can you please refer me to an implementation of multi part upload client?

@omar391 There's currently no implementation on the client side in ReactPHP for this, but it definitely sounds like a useful feature to add to this project. I didn't find any open tickets for this, so maybe it's worth to open up a [discussion](https://github.com/orgs/reactphp/discussions) inside our recently activated GitHub Discussions. This way we can talk about a potential implementation for this and also make this topic more visible for others!

**Comment by SimonFrings** [comment]
FYI: See https://github.com/orgs/reactphp/discussions/471 for further conversation.

**Pull request closed without merging**